File: slope_serial_threaded.py
import numpy as np
import threading
from multiprocessing import Process, Queue, Condition
from collections import deque
from time import sleep

"""
Threaded serial slope calculator for raster ascii files.
Uses producer-consumer paradigm to handle memory.

Charlie Kazer
"""

def main():

	input_file = open("aigrid.asc", 'r')
	ncols = input_file.readline().split()[1]
	nrows = input_file.readline().split()[1]
	xllcorner = input_file.readline().split()[1]
	yllcorner = input_file.readline().split()[1]
	cellsize = float(input_file.readline().split()[1])
	NODATA = float(input_file.readline().split()[1])


        #Create data buffer, conditions. Choose buffer size appropriate for memory size.
        data_buffer = Queue(maxsize=10000)

        output_file = open("output_slope.asc", 'w')
        
        #create threads
        load_proc = threading.Thread(target=load_func, args=(input_file, data_buffer))
        calc_proc = threading.Thread(target=calc_func, args=(output_file, data_buffer, int(nrows), int(ncols), float(cellsize), float(NODATA)))

        #set up header
        header_str = ("ncols %s\n"
                  "nrows %s\n"
                  "xllcorner %s\n"
                  "yllcorner %s\n"
                  "cellsize %f\n"
                  "NODATA_value %d\n"
                  % (ncols, nrows, xllcorner, yllcorner, cellsize, NODATA)
                 )

        output_file.write(header_str)

        #spin up header
        load_proc.start()
        calc_proc.start()

        calc_proc.join()

        #clean-up
        input_file.close()
        output_file.close()


#////////////////////////////////////////////////////////////////////////////////#

# load_func: Produces data from input file
def load_func(input_file, data_buffer):
  for line in input_file:
    # NOTE: Don't skip any lines here, the file pointer has already advanced
    # past the header to the data.
    while (data_buffer.full()):
      #So inefficient... But in order to get proper waiting, would need to introduce lock
      sleep(0.1)
      #print "loader waiting..."
      
    data_buffer.put(np.fromstring(line, sep=' '))


#////////////////////////////////////////////////////////////////////////////////#

# calc_func: Consumes data from shared buffer. Calls calc_slope to handle actual calculations
# Writes results to output file.
def calc_func(output_file, data_buffer, numRows, numCols, cellsize, NODATA):
  cur_lines = deque([], 3)
  count = 0
  cur_slope = []

  cur_lines.append(np.zeros(numCols).fill(NODATA))
  #Read first two lines so that when we enter main while, cur_lines will
  #always contain 3 lines
  while(len(cur_lines) < 2):
    while (data_buffer.empty()):
      sleep(0.1)
    cur_lines.append(data_buffer.get())
    count += 1

  #Calculate slope for top line
  for i in range(cur_lines[1].size):
    cur_slope.append(calc_slope(cur_lines, i, cellsize, NODATA))
  output_file.write(' '.join(cur_slope))
  output_file.write('\n')
  cur_slope = []

  #Main loop
  while (count < numRows):
    while (data_buffer.empty()):
      sleep(0.1)
      #print "waiting consumer..."

    cur_lines.append(data_buffer.get())
    count += 1

    for i in range(cur_lines[1].size):
      cur_slope.append(calc_slope(cur_lines, i, cellsize, NODATA))
    output_file.write(' '.join(cur_slope))
    output_file.write('\n')
    cur_slope = []

  #Calculate slope for bottom line
  cur_lines.append(np.zeros(numCols).fill(NODATA))
  for i in range(cur_lines[1].size):
      cur_slope.append(calc_slope(cur_lines, i, cellsize, NODATA))
  output_file.write(' '.join(cur_slope))
  output_file.write('\n')
  cur_slope = []



#////////////////////////////////////////////////////////////////////////////////#

# calc_slope: calculates the slope values for a cell of data
def calc_slope(cur_lines, col, cellsize, NODATA):
  if cur_lines[1][col] == NODATA:
    return str(NODATA)

  nbhd = [] #'neighborhood' of cell

  for i in range(3):
    for j in range(-1,2):
      if col+j<0 or col+j>=cur_lines[1].size:
        nbhd.append(NODATA)
      else:
        nbhd.append(cur_lines[i][col+j])

  dz_dx = (nbhd[2] + 2*nbhd[5] + nbhd[8] - (nbhd[0] + 2*nbhd[3] + nbhd[6])) \
                          / (8*cellsize)
  dz_dy = (nbhd[6] + 2*nbhd[7] + nbhd[8] - (nbhd[0] + 2*nbhd[1] + nbhd[2])) \
                            / (8*cellsize)

  slope = np.arctan(np.sqrt(np.square(dz_dx) + np.square(dz_dy)))

  return str(slope)

if __name__ == '__main__':
	main()

File: slope_serial_threaded.cpp
#include <iostream>
#include <fstream>
#include <stdlib.h>
#include <string>
#include <sstream>
#include <math.h>
#include <vector>
#include <algorithm>
#include <iterator>
#include <boost/thread.hpp>
#include <deque>

using namespace std;

//declaring global variables for # of columns, rows, cell size, and no data value
int NCOLS, NROWS;
double CELLSIZE, NODATA;

//Variables shared between threads
boost::condition_variable_any buffer_available;
boost::mutex buffer_lock;
deque<deque <double> > buffer;
const static int MAX_BUF_SIZE = 70000;

//function headers
void getHeader(ifstream* inFile, ofstream* outFile, string header[]);
void loadData(ifstream* inFile);
void calcFunc(ofstream* outFile);
double calc_slope(deque< deque <double> >* cur_lines, int col);



int main(int argc, char* argv[])
{
        if (argc != 3)
        {
                cerr << "Incorrect number of args" << endl;
                cout << "Correct usage: ./Slope <input filename> <output filename>" << endl;
                return 1;
        }
	
        string filename(argv[1]);
	string outputFile(argv[2]);
	string header[6];

        //Open files
        ifstream inFile;
        inFile.open(filename.c_str());
        if(!inFile.is_open())
        {
                cerr << "File failed to open" << endl;
                exit(1);
        }

	ofstream outFile;
	outFile.open(outputFile.c_str());
	if(!outFile.is_open())
	{
		cerr << "File failed to open" << endl;
		exit(1);
	}
	
        getHeader(&inFile, &outFile, header);

        //start threads
        boost::thread load_thread(loadData, &inFile);
        boost::thread calc_thread(calcFunc, &outFile);

        //wait for thread completion
        load_thread.join();
        calc_thread.join();

        inFile.close();
        outFile.close();

        return 0;
}

/* getHeader - Gathers header information from input file, and then
 *             writes that information to the output file since it will
 *             be identical.
 */
void getHeader(ifstream* inFile, ofstream* outFile, string header[])
{
	int count = 0;
	vector<string> keyValues;

	//reading lines of header and saving keyvalues into global variables
	while (count < 6)
	{
		getline(*inFile, header[count]);
		if(count == 0)
		{
			istringstream headLine(header[count]);			
			copy(istream_iterator<string>(headLine),
				istream_iterator<string>(),
				back_inserter(keyValues));
			NCOLS = atoi(keyValues[1].c_str());
			count++;
		}
		else if(count == 1)
                {
			istringstream headLine(header[count]);
                        copy(istream_iterator<string>(headLine),
                                istream_iterator<string>(),
                                back_inserter(keyValues));
                        NROWS = atoi(keyValues[3].c_str());
			count++;
                }
		else if(count == 4)
                {
			istringstream headLine(header[count]);
                        copy(istream_iterator<string>(headLine),
                                istream_iterator<string>(),
                                back_inserter(keyValues));
                        CELLSIZE = atof(keyValues[5].c_str());
			count++;
                }
		else if(count == 5)
                {
			istringstream headLine(header[count]);
                        copy(istream_iterator<string>(headLine),
                                istream_iterator<string>(),
                                back_inserter(keyValues));
                        NODATA = atof(keyValues[7].c_str());
			count++;
                }
		else
			count++;
	}

        //Write header to output file
	for (int i = 0; i < 6; i++)
	{
		*outFile << header[i] << '\n';
	}
}

/* loadData - Thread function which loads data into the shared buffer,
 *            line by line.
 */
void loadData(ifstream* inFile){
        string line;
        deque< double > row;

        //Read file line by line
        while(getline(*inFile, line))
        {
		//removing whitespace in the begining of the lines
                line.erase(line.begin());
		istringstream ss(line);
                string x;
                //Copy each token from a line into deque
                while(getline(ss, x, ' '))
		{
                        row.push_back(atof(x.c_str()));
          	}
                ///////////////LOCK////////////////////
                boost::mutex::scoped_lock lock(buffer_lock);
                while(buffer.size() == MAX_BUF_SIZE){
                    buffer_available.wait(buffer_lock);
                }
                buffer.push_back(row);
                buffer_available.notify_one();
                buffer_lock.unlock();
                /////////////UNLOCK///////////////////
                row.clear();
        }
}

/* calcFunc - Thread function which consumes data from the shared buffer,
 *            line by line. Also writes results of calculations to
 *            output file.
 */
void calcFunc(ofstream* outFile){
    deque< deque <double> >* cur_lines = new deque< deque <double> >;
    int count=0;
    int i;
    double cur_slope[NCOLS];

    //First push back NODATA row for calculating sloep of first row
    cur_lines->push_back(deque<double> (NCOLS, NODATA));
    //Next, grab first two rows of data
    for(i=0; i<2; i++){
        //////////////////////LOCK/////////////////////////
        boost::mutex::scoped_lock lock(buffer_lock);
        while(buffer.size() == 0){
            buffer_available.wait(buffer_lock);
        }
        //DONT pop anything from cur_lines yet, need to fill with three rows.
        cur_lines->push_back(buffer.front());
        buffer.pop_front();
        buffer_available.notify_one();
        buffer_lock.unlock();
        ////////////////////UNLOCK/////////////////////////
        count++;
    }
    //Calculate and write out first row
    for(i=0; i<NCOLS; i++){
        *outFile << calc_slope(cur_lines, i) << " ";
    }
    *outFile << "\n";

    //Enter main while loop
    while (count < NROWS){
        cur_lines->pop_front();
        //////////////////////LOCK/////////////////////////
        boost::mutex::scoped_lock lock(buffer_lock);
        while(buffer.size() == 0){
            buffer_available.wait(buffer_lock);
        }
        cur_lines->push_back(buffer.front());
        buffer.pop_front();
        buffer_available.notify_one();
        buffer_lock.unlock();
        ////////////////////UNLOCK/////////////////////////
        count++;
        for(i=0; i<NCOLS; i++){
            *outFile << calc_slope(cur_lines, i) << " ";
        }
        *outFile << "\n";
    }


    //Push back another NODATA row to calculate the last row with.
    cur_lines->pop_front();
    cur_lines->push_back(deque<double> (NCOLS, NODATA));
    //Calculate and write out last row
    for(i=0; i<NCOLS; i++){
        *outFile << calc_slope(cur_lines, i) << " ";
    }
    *outFile << "\n";

    delete cur_lines;
}

/* calc_slope - calculates the slope for a single cell in a raster
 *              file. cur_lines includes the input row of the cell, and
 *              the rows above and below that cell. col is the column
 *              of the cell we're calculating the slope for.
 */
double calc_slope(deque< deque <double> >* cur_lines, int col)
{
        if (cur_lines->at(1)[col] == NODATA){
            return NODATA;
        }

        double nbhd[9];//'neighborhood' of current cell
        int k=0;
        
        for (int i=0; i<3; i++){
            for (int j=-1; j<2; j++){
                if ((col+j < 0) or (col+j > NCOLS)){
                    nbhd[k] = NODATA;
                }
                else{
                    nbhd[k] = cur_lines->at(i)[col+j];
                }
                k++;
            }
        }

        double dz_dx = (nbhd[2] + (2*nbhd[5]) + nbhd[8] - (nbhd[0] + (2*nbhd[3]) + nbhd[6])) / (8*CELLSIZE);
        double dz_dy = (nbhd[6] + (2*nbhd[7]) + nbhd[8] - (nbhd[0] + (2*nbhd[1]) + nbhd[2])) / (8*CELLSIZE);

        return atan(sqrt(pow(dz_dx, 2) + pow(dz_dy, 2)));
}
